site/app/models/notebook/UserSpecificNotebook.php

```diff
@@ -0,0 +1,141 @@
+        );
+        $json = FileUtils::readJsonFile($tgt_dir);
+
+        if ($json !== false && isset($json['item_pool'])) {
+            $this->item_pool = $json['item_pool'];
+        }
+
+        $this->gradeable_id = $gradeable_id;
+        $this->user_id = $user_id;
+
+        $this->notebook_config = $this->replaceNotebookItemsWithQuestions($details);
+    }
+
+
+    //collect items from a notebook and replace them with the actual
+    //notebook values
+    private function replaceNotebookItemsWithQuestions($raw_notebook) {
+        $new_notebook = [];
+        $seen_items = [];
+        $tests = [];
+
+        foreach ($raw_notebook['notebook'] as $notebook_cell) {
+            if (isset($notebook_cell['type']) && $notebook_cell['type'] === 'item') {
+                //see if theres a target item pool and replace this with the actual notebook
+                $tgt_item = $this->getItemFromPool($notebook_cell);
+
+                $item_data = $this->searchForItemPool($tgt_item);
+                if (count($item_data['notebook']) > 0) {
+                    $new_notebook = array_merge($new_notebook, $item_data['notebook']);
+                    $tests = array_merge($tests, $item_data['testcases']);
+                }
+
+                //record if we potentially grabbed the same question
+                $seen_items = array_merge($seen_items, $notebook_cell['from_pool']);
+            }
+            else {
+                $new_notebook[] = $notebook_cell;
+            }
+        }
+        $this->test_cases = $tests;
+        $multiple_selected = [];
+        foreach (array_count_values($seen_items) as $seen => $value) {
+            if ($value > 1) {
+                $multiple_selected[] = $seen;
+            }
+        }
+
+        if (count($multiple_selected) > 0) {
+            $seen = implode($multiple_selected, ",");
+            $this->warning = "Warning: The item_pool \"{$seen}\" has been selected multiple times, this could lead to notebooks with the same question given more than once.";
+        }
+
+        return $new_notebook;
+    }
+
+
+    private function getItemFromPool($item) {
+        $item_label = $item['item_label'];
+        $selected = $this->getNotebookHash($item_label, count($item['from_pool']));
+        $item_from_pool = $item['from_pool'][$selected];
+        $this->selected_questions[] = $item_from_pool;
+
+        return $item_from_pool;
+    }
+
+    private function getNotebookHash($item_label, $from_pool_count) {
+    
+        $gid = $this->gradeable_id;
+        $uid = $this->user_id;
+
+        $semester = $this->core->getConfig()->getSemester();
+        $course = $this->core->getConfig()->getCourse();
+        
+        $hash = hexdec(substr(md5("{$item_label}|{$gid}|{$uid}|{$semester}|{$course}"), 24, 8));
+
+        $selected = $hash % $from_pool_count;
+        $this->hashes[] = $hash;
+
+        return $selected;
+    }
+
+
+    private function searchForItemPool($tgt_name) {
```

Comment:
```suggestion
    private function searchForItemPool(string $tgt_name): array {
```

Reply:
@shailpatels @MasterOdin 
Status of this requested change?